### PR TITLE
Read local version from `VERSION` file.

### PR DESCRIPTION
**Pull request by jdeanwallace**
Resolves https://github.com/tiny-pilot/tinypilot/issues/1021

Notes:
1. In local development mode (i.e. debug mode) we return [a dummy version hash `0000000`](https://github.com/tiny-pilot/tinypilot/blob/bb8c7f7999544961dba5e6e58149e940058393cb/app/version.py#L38).
2. In order to detect if we're in dev mode, we access `flask.current_app.debug`. This makes it notoriously difficult to write tests because we need to initialize a new Flask `app` to be able to give the unit test an application context. Anyway, instead of doing that I just pulled the `flask.current_app.debug` into its own function (i.e. [`_is_debug`](https://github.com/tiny-pilot/tinypilot/blob/bb8c7f7999544961dba5e6e58149e940058393cb/app/version.py#L21-L22)) and [mocked that instead](https://github.com/tiny-pilot/tinypilot/blob/bb8c7f7999544961dba5e6e58149e940058393cb/app/version_test.py#L12).

---
This change is [Reviewable](https://reviewable.io/reviews/tiny-pilot/tinypilot/1040)

### CODE REVIEW

**Review by mtlynch — CHANGES_REQUESTED** [review]
##### Automated comment from CodeApprove ([go to review](https://codeapprove.com/pr/tiny-pilot/tinypilot/1040))



<hr />

In: **app/version_test.py**:

```
> Line 40
        with tempfile.TemporaryDirectory() as mock_version_dir:
```
It feels like a lot of work to create a whole `NamedTemporaryDirectory` just to force the file not to exist. What about just specifying the filename as `non-existent-file.txt`?

<hr />

In: **app/version_test.py**:

```
> Line 13
            self.assertEqual('0000000', version.local_version())
```
I don't think it's worth testing debug mode. I'd say keep it if it was easy, but it prevents us from using a simple patcher (see below) to mock the debug behavior, so I think we can just cut it.

<hr />

In: **app/version_test.py**:

```
> Line 24
                                       return_value=False):
```
I like keeping as much as possible in the test function, but we're getting to so many mocks that it's a little hard to follow the test logic. What about extracting the `_is_debug` patch to a class-wide patch?

Suggested implementation: https://github.com/tiny-pilot/tinypilot/compare/local-version...local-version-debug-patch?expand=1

<hr />

In: **app/version_test.py**:

```
> Line 49
                    version.local_version()
```
It would be good to add a test for when the `VERSION` file does not contain valid UTF-8.

```python
    def test_local_version_raises_file_error_when_file_is_not_utf_8(self):
        with tempfile.NamedTemporaryFile('wb') as mock_version_file:
            mock_version_file.write(bytearray(b'\xff'))
            mock_version_file.flush()

            with mock.patch.object(version, '_VERSION_FILE',
                                   mock_version_file.name):
                with self.assertRaises(version.FileError):
                    version.local_version()
```

<hr />

In: **app/version_test.py**:

```
> Line 48
                with self.assertRaises(version.FileError):
```
Can we add a test for when the VERSION file is empty? We should raise a `FileError` in that case.

<hr />

In: **app/version.py**:

```
> Line 14
class FileError(Error):
```
In Python, the module name doesn't appear when the error is thrown, so when this happens, we'll just see:

```text
 FileError: Failed to check local version: blah blah
```

`FileError` seems too much like a Python stdlib name. What about `VersionFileError`?


<hr />

👀 @jdeanwallace it's your turn please take a look

**Review by jdeanwallace — COMMENTED** [review]
##### Automated comment from CodeApprove ([go to review](https://codeapprove.com/pr/tiny-pilot/tinypilot/1040))



<hr />

In: **app/version_test.py**:
Ah good idea. Thanks!

<hr />

In: **app/version_test.py**:

```
> Line 21
            self.assertEqual('0000000', version.local_version())
```
Thanks! I managed to keep this test after applying the suggested class-wide patch by just re-patching `_is_debug` to return `True`.

<hr />

In: **app/version_test.py**:
Haha that works. Yeah I was overthinking it and wanted the file path to be random. Thanks!

<hr />

In: **app/version_test.py**:
Done.

<hr />

In: **app/version_test.py**:

```
> Line 66
                    version.local_version()
```
Done.

<hr />

In: **app/version.py**:
Done.


<hr />

👀 @mtlynch it's your turn please take a look

**Review by mtlynch — CHANGES_REQUESTED** [review]
##### Automated comment from CodeApprove ([go to review](https://codeapprove.com/pr/tiny-pilot/tinypilot/1040))


✔️ Approved (1 unresolved comments)

LGTM

<hr />

In: **app/version_test.py**:

```
> Line 21
            self.assertEqual('0000000', version.local_version())
```
Double patching is kind of confusing. Can we pull it out of the `VersionTest` class and put it in its own test class without the `setUp`?


<hr />

👀 @jdeanwallace it's your turn please take a look

**Review by jdeanwallace — COMMENTED** [review]
##### Automated comment from CodeApprove ([go to review](https://codeapprove.com/pr/tiny-pilot/tinypilot/1040))



<hr />

In: **app/version_test.py**:
Done.

**Review by mtlynch — APPROVED** [review]
##### Automated comment from CodeApprove ([go to review](https://codeapprove.com/pr/tiny-pilot/tinypilot/1040))

Approved: I have approved this change on CodeApprove and all of my comments have been resolved.
